src/frontend/src/flows/authenticate/index.ts

```diff
@@ -0,0 +1,390 @@
+          resolve(authSuccess);
+        }
+      });
+    };
+  });
+};
+
+const authenticateUser = async (
+  authContext: AuthContext,
+  pageMode: PageMode
+): Promise<AuthSuccess | null> => {
+  try {
+    const userNumber = readUserNumber(pageMode);
+    if (userNumber === undefined) {
+      toggleErrorMessage("userNumberInput", "invalidAnchorMessage", true);
+      return null;
+    }
+    const result = await withLoader(() => IIConnection.login(userNumber));
+    const loginResult = apiResultToLoginFlowResult(result);
+    return await handleAuthResult(loginResult, authContext);
+  } catch (error) {
+    await displayError({
+      title: "Authentication Failed",
+      message: "An error occurred during authentication.",
+      detail: error instanceof Error ? error.message : JSON.stringify(error),
+      primaryButton: "Try again",
+    });
+    return init(authContext);
+  }
+};
+
+const displayPage = (
+  origin: string,
+  maxTimeToLive: bigint | undefined,
+  userNumber: bigint | undefined
+) => {
+  const container = document.getElementById("pageContent") as HTMLElement;
+  render(pageContent(origin, maxTimeToLive, userNumber), container);
+  const pageMode: PageMode = {
+    mode: userNumber === undefined ? "newUserNumber" : "existingUserNumber",
+  };
+  setMode(pageMode);
+  return pageMode;
+};
+
+/**
+ * Checks whether login was successful, and if so fetches the delegation from II.
+ * @param loginResult Result of authentication
+ * @param authContext Information about the authentication request being processed.
+ */
+const handleAuthResult = async (
+  loginResult: LoginFlowResult,
+  authContext: AuthContext
+): Promise<AuthSuccess> => {
+  if (loginResult.tag === "ok") {
+    // successful login, store user number for next time
+    setUserNumber(loginResult.userNumber);
+    const [userKey, parsed_signed_delegation] = await withLoader(() =>
+      fetchDelegation(loginResult, authContext)
+    );
+    return new AuthSuccess(loginResult.userNumber, loginResult.connection, () =>
+      authContext.postMessageCallback({
+        kind: "authorize-client-success",
+        delegations: [parsed_signed_delegation],
+        userPublicKey: Uint8Array.from(userKey),
+      })
+    );
+  } else {
+    await displayError({
+      title: loginResult.title,
+      message: loginResult.message,
+      detail: loginResult.detail !== "" ? loginResult.detail : undefined,
+      primaryButton: "Try again",
+    });
+    return init(authContext);
+  }
+};
+
+const initRecovery = (pageMode: PageMode) => {
+  const recoverButton = document.getElementById(
+    "recoverButton"
+  ) as HTMLButtonElement;
+  recoverButton.onclick = () => {
+    const userNumber = readUserNumber(pageMode);
+    if (userNumber !== undefined) {
+      return useRecovery(userNumber);
+    }
+    toggleErrorMessage("userNumberInput", "invalidAnchorMessage", true);
+  };
+};
+
+type PageMode = { mode: "existingUserNumber" | "newUserNumber" };
```

Comment:
```suggestion
type PageMode = "existingUserNumber" | "newUserNumber";
```

---

EDIT: actually reading the code, I suspect you're relying on the fact that `mode` is a reference

Reply:
Yes, exactly. Is this ok?